src/ApiService/ApiService/Program.cs

```diff
@@ -58,6 +58,67 @@ public class LoggingMiddleware : IFunctionsWorkerMiddleware {
         }
     }
 
+    /// <summary>
+    /// Represents a middleware that can optionally perform strict version checking based on data sent in request headers.
+    /// </summary>
+    public class VersionCheckingMiddleware : IFunctionsWorkerMiddleware {
+        private const string CliVersionHeader = "Cli-Version";
+        private const string StrictVersionHeader = "Strict-Version";
+        private readonly Version _oneFuzzServiceVersion;
+        private readonly IRequestHandling _requestHandling;
+
+        /// <summary>
+        /// Initializes an instance of <see cref="VersionCheckingMiddleware"/> with the provided config and request handling objects.
+        /// </summary>
+        /// <param name="config">The service config containing the service version.</param>
+        /// <param name="requestHandling">The request handling object to create HTTP responses with.</param>
+        public VersionCheckingMiddleware(IServiceConfig config, IRequestHandling requestHandling) {
+            _oneFuzzServiceVersion = Version.Parse(config.OneFuzzVersion);
+            _requestHandling = requestHandling;
+        }
+
+        public Error? CheckCliVersion(Azure.Functions.Worker.Http.HttpHeadersCollection headers) {
```

Comment:
This could be a `OnefuzzResultVoid` so we don't have to encode null as success, we can return `Ok` instead.